Feature/2470 bucket schema (#2765)
influxdata/docs-v2: content/influxdb/cloud/organizations/bucket-schema/_index.md

    ```

2. Use the [`influx bucket-schema list` command](/influxdb/cloud/reference/cli/influx/bucket-schema/list) to list bucket-schemas.
   To output schema column details, use the optional `extended-output` flag.

    ```sh
    influx bucket-schema list --bucket my_explicit_bucket --extended-output
    ```

### Update a schema

  Use the [`influx bucket-schema update` command](/influxdb/cloud/reference/cli/influx/bucket-schema/update) to add new columns to a schema. You cannot modify or delete columns from your schema.

### Errors

#### Not permitted by schema
  If the measurement data doesn't conform to the bucket schema for that measurement, InfluxDB returns an error.

  The following example attempts to write a *cpu* measurement with an incorrect `usage_user` [data type](/influxdb/cloud/reference/glossary/#data-type):

  ```sh
  influx write -b my_explicit_bucket 'cpu,host=myHost usage_user="1001" 1556896326'
  ```

  The output is the following:
  ```sh
  Error: failed to write data:
  unable to parse 'cpu,host=myHost usage_user="1001" 1556896326':
  schema: field type for field "usage_user" not permitted by schema; got String but expected Float

>   ```

Reviewer: No need to indent these paragraphs.

~~~suggestion
Use the [`influx bucket-schema update` command](/influxdb/cloud/reference/cli/influx/bucket-schema/update) to add new columns to a schema. You cannot modify or delete columns from your schema.

### Errors

#### Not permitted by schema
If the measurement data doesn't conform to the bucket schema for that measurement, InfluxDB returns an error.

The following example attempts to write a *cpu* measurement with an incorrect `usage_user` [data type](/influxdb/cloud/reference/glossary/#data-type):

```sh
influx write -b my_explicit_bucket 'cpu,host=myHost usage_user="1001" 1556896326'
```

The output is the following:
```sh
Error: failed to write data:
unable to parse 'cpu,host=myHost usage_user="1001" 1556896326':
schema: field type for field "usage_user" not permitted by schema; got String but expected Float
```
~~~

#### No measurement schemas
If you attempt to write to a bucket that has schema-type `explicit` but doesn't have a schema, the
bucket rejects write attempts.

  The output is the following:

  ```sh
  Error: failed to write data: schema: bucket "my_explicit_bucket" contains
  no measurement schemas
  ```

#### Failed to create measurement
  If you attempt to add a new schema with the
  same measurement name as an existing schema, InfluxDB rejects the new schema.

  The output is the following:
  ```sh
  Error: failed to create measurement: 422 Unprocessable Entity

>   ```

Reviewer: Same here. No need for indentation.

~~~suggestion
The output is the following:

```sh
Error: failed to write data: schema: bucket "my_explicit_bucket" contains
no measurement schemas
```

#### Failed to create measurement
If you attempt to add a new schema with the
same measurement name as an existing schema, InfluxDB rejects the new schema.

The output is the following:
```sh
Error: failed to create measurement: 422 Unprocessable Entity
```
~~~

---
title: Manage bucket schemas
seotitle: Manage bucket schemas in InfluxDB
description: Manage bucket schemas in InfluxDB using the influx CLI.
menu:
  influxdb_cloud:
    name: Manage bucket schemas
    parent: Manage organizations
weight: 105
influxdb/cloud/tags: [buckets, bucket-schema, schema]
---

Use *bucket schemas* to restrict a bucket's columns and data types.

A **bucket schema** enforces the shape of data in a bucket.
By default, buckets have an `implicit` **schema-type**  which conforms
to data as it is written to InfluxDB.

Buckets with the `explicit` schema-type allow you to assign your own named schemas
and reject write requests that do not conform to a specific schema.

### Create a bucket schema
Use the `influx` CLI to set the schema-type and one or more schemas for your bucket:

> 1. Create a measurement schema *columns file* using CSV, JSON, or [Newline delimited JSON (NDJSON)](http://ndjson.org/). Define the name, type, and data type of each column.

Author: Good idea. In this context, they're effectively the same. bucket-schema is essentially a join of a measurement schema to a bucket. I was trying to be careful in case measurement schema is used in other contexts (now or in the future).

> 1. Create a measurement schema *columns file* using CSV, JSON, or [Newline delimited JSON (NDJSON)](http://ndjson.org/). Define the name, type, and data type of each column.

Author: ...also, "measurement name" is used to distinguish schema.

  ```

  {{% /code-tab-content %}}
  {{% code-tab-content %}}

  ```json
  {{< get-assets-text "bucket-schema/bucket-schema-columns.json" >}}
  ```

  {{% /code-tab-content %}}
  {{% code-tab-content %}}

  ```json
  {{< get-assets-text "bucket-schema/bucket-schema-columns.ndjson" >}}
  ```

  {{% /code-tab-content %}}
  {{< /code-tabs-wrapper >}}

2. Create a bucket with the `schema-type` flag set to `explicit`.

    ```sh
    {{< get-assets-text "bucket-schema/bucket-schema-type.sh" >}}
    ```

    To view your bucket's schema-type, use the [`influx bucket list` command](/influxdb/cloud/reference/cli/influx/bucket-schema/list).

    With schema-type `explicit`, but no bucket-schema, the
    bucket rejects write attempts.

    ```sh
    influx write --bucket my_explicit_bucket --format=lp "cpu,host=foo usage_user=0.5"
    ```

    The output is the following:
    ```sh
    Error: failed to write data: schema: bucket "my_explicit_bucket" contains
    no measurement schemas
    ```

3. Use the [`influx bucket-schema create` command](/influxdb/cloud/reference/cli/influx/bucket-schema/create) to assign the schema to your bucket.

>    Provide the location of your schema columns file with the `columns-file` flag and the name of your measurement with the `name` flag.

Author: Each schema only has a `measurement name` set by flag `--name`. The data's `_measurement` value will be matched to the schema's measurement name.

   To output schema column details, use the optional `extended-output` flag.

   ```sh
   influx bucket-schema create \
     --bucket my_explicit_bucket \
     --name cpu \
     --columns-file columns.csv \
     --extended-output
   ```

   InfluxDB responds with the created schema:

   ```sh
   ID                      Measurement Name       Column Name     Column Type   Column Data Type    Bucket ID
   07b80ed3bb73e000        cpu                    time            timestamp     4d5dbdfd5a67ae4f
   07b80ed3bb73e000        cpu                    host            tag           4d5dbdfd5a67ae4f
   07b80ed3bb73e000        cpu                    usage_user      field         float               4d5dbdfd5a67ae4f
   ```

   #### Columns format

    By default, InfluxDB attempts to detect the columns file format.

>     To specify the format, use the [`columns-format` flag](/influxdb/cloud/reference/cli/influx/bucket-schema/create).

Author: > Does this mean if you want to specify a columns file format that is different that the format of the saved schema file?

I tested a little bit. The `auto`-detection just infers the format by the file extension. If your file is named "my-schema.txt" and contains JSON then you have to specify `columns-format json`.

schemas to a bucket.

{{% note %}}
{{< bucket-schema-type >}}
```sh
{{< get-assets-text "bucket-schema/bucket-schema-type.sh" >}}
```
{{% /note %}}

1. Add a new schema with a new measurement name.

    ```sh
    influx bucket-schema create \
      --bucket my_explicit_bucket \
      --name temperature \
      --columns-file schema.json
    ```

    InfluxDB responds with an error if you attempt to add a new schema with the
    same measurement name as an existing schema, as in the following example:

    ```sh
    influx bucket-schema create \
      --bucket my_explicit_bucket \
      --name cpu \
      --columns-file schema.json
    ```

    The output is the following:
    ```sh
    Error: failed to create measurement: 422 Unprocessable Entity
    ```

2. Use the [`influx bucket-schema list` command](/influxdb/cloud/reference/cli/influx/bucket-schema/list) to list bucket-schemas.
   To output schema column details, use the optional `extended-output` flag.

    ```sh
    influx bucket-schema list --bucket my_explicit_bucket --extended-output
    ```
### Update a measurement schema

>   Use the [`influx bucket-schema update` command](/influxdb/cloud/reference/cli/influx/bucket-schema/update) to update a bucket's measurement schema.

Author: Not sure. That would mean losing some data, I assume.

---
title: Manage bucket schema
seotitle: Manage bucket schema in InfluxDB
description: Manage bucket schema in InfluxDB using the influx CLI.
menu:
  influxdb_cloud:
    name: Manage bucket schema
    parent: Manage organizations
weight: 105
influxdb/cloud/tags: [buckets, bucket-schema, schema]
---

Use *bucket-schema* to restrict a bucket's columns and data types.

*Bucket-schema* are measurement schema that enforce the shape of
your bucket data.
By default, buckets receive *schema-type* `implicit` and InfluxDB's default
measurement schema.

> Buckets with schema-type `explicit` allow you to assign your own schema.
> These buckets reject write requests that do not obey their bucket-schema.

Reviewer: ```suggestion
Buckets with an `explicit` schema-type let you specify and enforce your own schema.
InfluxDB rejects write requests to buckets with an `explicit` schema-type that do not conform to the specified bucket schema.
```

Author: @sanderson Thanks - rewrote this intro ~20 different ways.

### Create a bucket schema

> Use the influx CLI to set the schema-type and one or more schema for your bucket:

Author: Oops - schema***s*** never seems right to me.

> Use the influx CLI to set the schema-type and one or more schema for your bucket:

Reviewer: I had to double check as well 😄. "Schemata" or "schemas" are the plural form of "schema."

1. Create a bucket with the `schema-type` flag set to `explicit`.

   ```sh
   {{< get-assets-text "bucket-schema/bucket-schema-type.sh" >}}
   ```

   To view your bucket's schema-type, use the [`influx bucket list` command](/influxdb/cloud/reference/cli/influx/bucket-schema/list).

   With schema-type `explicit` but no bucket-schema, the
   bucket rejects write attempts.

   For example,
   ```sh
   influx write --bucket my_explicit_bucket --format=lp "cpu,host=foo usage_user=0.5"
   ```

   returns the following error:

   ```sh
   Error: failed to write data: schema: bucket "my_explicit_bucket" contains
   no measurement schemas

>    ```

Reviewer: I'd move this down after you walkthrough adding a schema.

> 2. Create a measurement schema *columns file* using CSV, JSON, or Newline delimited JSON (NDJSON).
>    Define the name, field type, and data type of each column.

Reviewer: ```suggestion
2. Create a measurement schema *columns file* using CSV, JSON, or [Newline delimited JSON (NDJSON)](http://ndjson.org/).
   Define the name, type, and data type of each column.
```

We should probably also list the supported column `types` and `datatypes`:

```
- **types**
    - timestamp
    - field
    - tag
- **data types** (only field columns require a dataType)
    - string
    - bool
    - float
    - int
    - uint
 ```

>    {{< code-tabs-wrapper >}}
>    {{% code-tabs %}}
>    [columns.csv](#)
>    [columns.json](#)
>    [columns.ndjson](#)
>    {{% /code-tabs %}}
>    {{% code-tab-content %}}
>    ```sh
>    {{< get-assets-text "bucket-schema/bucket-schema-columns.csv" >}}
>    ```
>    {{% /code-tab-content %}}
>    {{% code-tab-content %}}
>    ```json
>    {{< get-assets-text "bucket-schema/bucket-schema-columns.json" >}}
>    ```
>    {{% /code-tab-content %}}
>    {{% code-tab-content %}}
>    ```json
>    {{< get-assets-text "bucket-schema/bucket-schema-columns.ndjson" >}}
>    ```
>    {{% /code-tab-content %}}
>    {{< /code-tabs-wrapper >}}

Reviewer: To get these to format correctly under the line item and not break the list numbering:

```suggestion
    {{< code-tabs-wrapper >}}
{{% code-tabs %}}
[columns.csv](#)
[columns.json](#)
[columns.ndjson](#)
{{% /code-tabs %}}
{{% code-tab-content %}}
```sh
{{< get-assets-text "bucket-schema/bucket-schema-columns.csv" >}}
```
{{% /code-tab-content %}}
{{% code-tab-content %}}
```json
{{< get-assets-text "bucket-schema/bucket-schema-columns.json" >}}
```
{{% /code-tab-content %}}
{{% code-tab-content %}}
```json
{{< get-assets-text "bucket-schema/bucket-schema-columns.ndjson" >}}
```
{{% /code-tab-content %}}
    {{< /code-tabs-wrapper >}}
```

> 3. Use the [`influx bucket-schema create` command](/influxdb/cloud/reference/cli/influx/bucket-schema/create) to set the schema for your bucket.
>    Pass the location of your columns file to the `columns-file` flag and the
>    name of your measurement schema to the `name` flag.

Reviewer: ```suggestion
3. Use the [`influx bucket-schema create` command](/influxdb/cloud/reference/cli/influx/bucket-schema/create) to set the schema for your bucket.
   Provide the location of your columns file with the `columns-file` flag and the
   name of your measurement with the `name` flag.
```

   ```sh
   influx bucket-schema create \
     --bucket my_explicit_bucket --name cpu \
     --columns-file columns.csv
     -x
   ```

   InfluxDB responds with the created schema.

   ```sh
   ID                      Measurement Name       Column Name     Column Type   Column Data Type    Bucket ID
   07b80ed3bb73e000        cpu                    time            timestamp     4d5dbdfd5a67ae4f
   07b80ed3bb73e000        cpu                    host            tag           4d5dbdfd5a67ae4f
   07b80ed3bb73e000        cpu                    usage_user      field         float               4d5dbdfd5a67ae4f
   ```

   #### Columns format

>    Use the [`columns-format` flag](/influxdb/cloud/reference/cli/influx/bucket-schema/create)
>    when you need to specify the format of your file.
>    InfluxDB defaults to `auto` and attempts to guess the format.

Reviewer: ```suggestion
    By default, InfluxDB attempts to detect the column file format.
    To specify the column file formation, use the [`columns-format` flag](/influxdb/cloud/reference/cli/influx/bucket-schema/create).  
```

### Add measurement schema to a bucket
Use the [`influx bucket-schema create` command](/influxdb/cloud/reference/cli/influx/bucket-schema/create) to add additional measurement
schema to a bucket.

1. Add a new schema with a new measurement name.

    ```sh
    influx bucket-schema create \
      -n my_explicit_bucket \
      --name temperature \
      --columns-file schema.json
    ```

    InfluxDB returns the new bucket-schema.

    ```sh
    ID                      Measurement Name        Bucket ID
    07b81b87c6deb000        temperature             4d5dbdfd5a67ae4f
    ```

2. Use the [`influx bucket-schema list` command](/influxdb/cloud/reference/cli/influx/bucket-schema/list) to list bucket-schemas. Your new schema should appear in the list.

    ```sh
    influx bucket-schema list -n my_explicit_bucket
    ```

    ```sh
    ### Output
    ID                      Measurement Name        Bucket ID
    07b81b87c6deb000        cpu                     4d5dbdfd5a67ae4f
    07b80ed3bb73e000        temperature             4d5dbdfd5a67ae4f

>     ```

Author: > 

I know it looks better, but I have mixed feelings :)

>     ```

Reviewer: This isn't a hill I'm going to die on 😄 . I guess my goal is to merge the two codeblocks into a single codeblock.

InfluxDB responds with an error if you attempt to add a new schema with the
same measurement name as an existing schema.

For example,
```sh
influx bucket-schema create \
  -n my_explicit_bucket \
  --name cpu \
  --columns-file schema.json
```

returns the error
```sh
Error: failed to create measurement: 422 Unprocessable Entity
```

### Write to the bucket

  Use the [`influx write` command](/influxdb/cloud/reference/cli/influx/write)
  to write data to a bucket.

  `write` returns an error
  if the bucket has schema-type `explicit` and the data does not obey a defined
   bucket-schema.

     ```sh
     influx write -b my_explicit_bucket 'cpu,host=myHost usage_user="1001" 1556896326'
     Error: failed to write data:
      unable to parse 'cpu,host=myHost usage_user="1001" 1556896326':
       schema: field type for field "usage_user" not permitted by schema; got String but expected Float

>      ```

Reviewer: ~~~suggestion
### Write to the bucket

Use the [`influx write` command](/influxdb/cloud/reference/cli/influx/write)
to write data to a bucket.

If the data does conform to the defined bucket-schema, InfluxDB returns an error:

```sh
$ influx write -b my_explicit_bucket 'cpu,host=myHost usage_user="1001" 1556896326'
Error: failed to write data:
unable to parse 'cpu,host=myHost usage_user="1001" 1556896326':
schema: field type for field "usage_user" not permitted by schema; got String but expected Float
```
~~~

---
title: Manage bucket schemas
seotitle: Manage bucket schemas in InfluxDB
description: Manage bucket schemas in InfluxDB using the influx CLI.
menu:
  influxdb_cloud:
    name: Manage bucket schemas
    parent: Manage organizations
weight: 105
influxdb/cloud/tags: [buckets, bucket-schema, schema]
related:
  - /influxdb/cloud/reference/key-concepts/
  - /influxdb/cloud/reference/key-concepts/data-schema/
  - /influxdb/cloud/reference/key-concepts/data-elements/
---

Use **bucket schemas** to ensure data includes specific [columns](/influxdb/cloud/reference/glossary/#column), [fields](/influxdb/cloud/reference/glossary/#field), and
 [data types](/influxdb/cloud/reference/glossary/#data-type).

{{% note %}}
#### Before you begin

The bucket schema examples below reference [**InfluxDB data elements**](/influxdb/cloud/reference/key-concepts/data-elements/).

> For more information, see [InfluxDB key concepts](/influxdb/cloud/reference/key-concepts/)

Reviewer: ```suggestion
For more information, see [InfluxDB key concepts](/influxdb/cloud/reference/key-concepts/).
```
